### PR TITLE
Use FFmpeg for Video Dumping instead of VFW in Windows

### PR DESCRIPTION
**Pull request by RisingFog**
Reduced the file sizes greatly by custom building ffmpeg to only include what is needed.

### CODE REVIEW

**Comment by Buddybenj** [comment]
Bump.

**Comment by Buddybenj** [comment]
This has merge conflicts now.

**Comment by skidau** [comment]
Please rebase this PR.  I would like to try it out.

File: Source/Core/VideoCommon/VideoCommon.vcxproj [hunk 1]
```diff
@@ -34,6 +34,26 @@
     <Import Project="..\..\VSProps\PCHUse.props" />
   </ImportGroup>
   <PropertyGroup Label="UserMacros" />
+  <ItemDefinitionGroup Condition="'$(Configuration)|$(Platform)'=='Debug|x64'">
```
Comment (shuffle2) [inline Source/Core/VideoCommon/VideoCommon.vcxproj]: I think you are doing this wrong.

**Comment by shuffle2** [comment]
Uh, so you built a custom ffmpeg and added the binaries to dolphin's repo?

**Comment by RisingFog** [comment]
Now that we've relicensed, does it make more sense to include ffmpeg by default in some form (DLL files or otherwise)?

**Comment by RisingFog** [comment]
Latest update now has pre-built static libraries now compiled into the executable file. Total size difference is just over 1mb from previous executable files.

**Comment by RisingFog** [comment]
Any major thing blocking this?

File: Source/Core/VideoCommon/VideoCommon.vcxproj (hunk 1 above)
Comment (RisingFog) [inline Source/Core/VideoCommon/VideoCommon.vcxproj]: which part is wrong?

**Comment by RisingFog** [comment]
@shuffle2 yes, these are pre-built static libraries.

**Comment by NewLunarFire** [comment]
Are we already using libav to dump video in other OS? I think libav/ffmpeg is the way to go for this, but I can't see what we currently use for other distros.

I also like the container and codec format change from AVI to MKV Matroska, but shouldn't we rename AVIDump(.cpp|.h) to MKVDump(.cpp|.h) then?

**Comment by RisingFog** [comment]
@NewLunarFire On Linux, it's only used if the builder has the pre-requisites for ffmpeg/libav. This PR is specifically for Windows, which doesn't really have a native way to build ffmpeg. For this, we are providing pre-built static libraries which will compile within the executable. I'm pretty sure that OSX will have to eventually have a PR created for the same purpose, as it currently doesn't work at all in OSX.

As for the name of the file itself, I don't think it's a big deal if we change the name of it.

**Comment by JosJuice** [comment]
If those file names are going to be changed, simply VideoDump is better in my opinion.

**Comment by NewLunarFire** [comment]
I'm not sure is naming that file VideoDump is a good idea. What will we do if we implement dumping to different container formats like OGG in the future? Should we put them in the same class and have a variable determine the format, or create separate MKVDump and OGGDump classes?

**Comment by RisingFog** [comment]
@NewLunarFire OGG is an audio format, not a video format

**Comment by JosJuice** [comment]
No, Ogg is a container format. Files with an .ogg extension are typically the Vorbis audio format in an Ogg container.

**Comment by NewLunarFire** [comment]
OGG is a container format, even thought Xiph.org said to use .ogg for audio and .ogv for video. And it supports a variety of codecs, and could probably hold H264 video with AAC audio.

That said, you probably removed Vorbis and Theora from your minimal FFMPEG build, which probably means we'll have to rebuild it with those codecs included to use the OGG format in the future.

**Comment by GameRecover** [comment]
What are the chances of this making it in to 5.0?
It's certainly a fix I'm excited for.

**Comment by RisingFog** [comment]
Reverting all additional changes past equivalent support to Linux to have this change available for 5.0

**Comment by RisingFog** [comment]
This can be reviewed now, the attempt to create a submodule failed with the buildbot. I'd prefer if this was merged before feature freeze.

**Comment by ZephyrSurfer** [comment]
Can this be updated to  new ffmpeg?
Since this was made they use hardware accelarted encoders in intel and nVidia (QSV & NVENC).

Maybe this is just a nicety, but I think video recording is demanding and this request would help somewhat.

**Comment by RisingFog** [comment]
@ZephyrSurfer Current version being used is 2.7.2 (same base as ppsspp-ffmpeg)

**Comment by vgturtle127** [comment]
Isn't 2.8.4 the latest? Why couldn't Dolphin and PPSSPP use that? Just curious.

**Comment by RisingFog** [comment]
@vgturtle127 it is, but the differences in ffmpeg versions for what we need it for does not really make it necessary to update

**Comment by vgturtle127** [comment]
@RisingFog Alright. I always keep my libraries and external stuff updated regardless, but that is because I have OCD about such things. :smile:

**Comment by RisingFog** [comment]
I've decided to go ahead and update to latest git for ffmpeg.

**Comment by ZephyrSurfer** [comment]
This seems to have a few issues

If framedump0avi exists it will overwrite it without asking unlike master.

**Comment by RisingFog** [comment]
@ZephyrSurfer This is only to bring the functionality up to par with Linux, which does the same thing. I'll see what I can do though.

**Comment by ZephyrSurfer** [comment]
I like that master can make lossless frame dumps without having me unable to keep full speed in the emulator.
So I think this pull request is inferior in that regard.

I don't see the need for linux to have the exact same options...

**Comment by RisingFog** [comment]
@ZephyrSurfer The problem with the current method is that it doesn't dump everything properly, and loses sync with audio over time. the VFW API (Video for Windows) is heavily outdated, and does not handle variable framerate content well like 3D video games.

**Comment by RisingFog** [comment]
@ZephyrSurfer latest push should fix the file overwriting without asking.

**Comment by ZephyrSurfer** [comment]
It overwrites anyway whether I select yes or no.

Plus if I select yes, as in overwrite, the time stamp of when the file was made does not change.

**Comment by RisingFog** [comment]
@ZephyrSurfer I actually just tried this on latest master and it still overwrites the video file even if I click no. Possibly something that never worked?

**Comment by RisingFog** [comment]
closing this PR until I can find a better solution which doesn't crash the buildbots

**Pull request closed without merging**